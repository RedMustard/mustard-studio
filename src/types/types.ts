export type KeyboardKeyCount = 88 | 61 | 49 | 37 | 25;
export type KeyboardKeyColor = 'white' | 'black';
export type Octave = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7;
export type OscillatorId = 'osc1' | 'osc2' | 'oscSub';
export type ValueUnit = 'percent' | 'cent' | 'pan';


export type StudioService = {
    settings: {
        master: {
            volume: number,
            pan: number,
        },
        osc1: OscillatorSettings & OscillatorDetune,
        osc2: OscillatorSettings & OscillatorDetune,
        oscSub: OscillatorSettings,
    },
    gainNodes: {
        master: GainNode,
        osc1: GainNode,
        osc2: GainNode,
        oscSub: GainNode,
    },
    panNodes: {
        master: StereoPannerNode,
        osc1: StereoPannerNode,
        osc2: StereoPannerNode,
        oscSub: StereoPannerNode,
    },
};

export type OscillatorSettings = {
    enabled: boolean,
    octave: number,
    pan: number,
    type: OscillatorType,
    volume: number,
};

export type OscillatorDetune = {
    detune: number,
};

export type StudioServiceActionType =
    'SET_MASTER_VOLUME' |
    'SET_MASTER_GAIN' |
    'SET_MASTER_PAN_NODE' |
    'SET_MASTER_PAN_POSITION' |
    'SET_OSC_1_VOLUME' |
    'SET_OSC_2_VOLUME' |
    'SET_OSC_SUB_VOLUME' |
    'SET_OSC_1_GAIN' |
    'SET_OSC_2_GAIN' |
    'SET_OSC_SUB_GAIN' |
    'SET_OSC_1_ENABLED' |
    'SET_OSC_2_ENABLED' |
    'SET_OSC_SUB_ENABLED' |
    'SET_OSC_1_TYPE' |
    'SET_OSC_2_TYPE' |
    'SET_OSC_SUB_TYPE' |
    'SET_OSC_1_DETUNE' |
    'SET_OSC_2_DETUNE' |
    'SET_OSC_1_PAN_NODE' |
    'SET_OSC_2_PAN_NODE' |
    'SET_OSC_SUB_PAN_NODE' |
    'SET_OSC_1_PAN_POSITION' |
    'SET_OSC_2_PAN_POSITION' |
    'SET_OSC_SUB_PAN_POSITION' |
    'SET_OSC_1_OCTAVE' |
    'SET_OSC_2_OCTAVE' |
    'SET_OSC_SUB_OCTAVE';

export type StudioServiceAction = {
    type: StudioServiceActionType;
    payload: any;
};

export type DispatchFunction = (args: StudioServiceAction) => void;

