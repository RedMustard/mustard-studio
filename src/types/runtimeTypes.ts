export enum OscillatorTypeEnum {
    SINE = 'sine',
    SAWTOOTH = 'sawtooth',
    TRIANGLE = 'triangle',
    SQUARE = 'square',
    CUSTOM = 'custom',
}
export const OscillatorTypes = Object.values(OscillatorTypeEnum) as OscillatorType[];

export enum FilterTypeEnum {
    LOWPASS = 'lowpass',
    HIGHPASS = 'highpass',
    BANDPASS = 'bandpass',
    LOWSHELF = 'lowshelf',
    HIGHSHELF = 'highshelf',
    PEAKING = 'peaking',
    NOTCH = 'notch',
    ALLPASS = 'allpass',
}
export const FilterTypes = Object.values(FilterTypeEnum) as BiquadFilterType[];


export enum ValueUnitEnum {
    PERCENT = 'percent',
    CENT = 'cent',
    PAN = 'pan',
    SECONDS = 'seconds',
}

export enum OscillatorIdEnum {
    OSC_1 = 'osc1',
    OSC_2 = 'osc2',
    OSC_SUB = 'oscSub',
}
