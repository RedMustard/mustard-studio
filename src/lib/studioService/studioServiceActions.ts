import { DispatchFunction, OscillatorId } from '../../types/types';
import { logger } from '../utils/logger/logger';

export const setMasterVolume = (value: number, dispatch: DispatchFunction) => {
    dispatch({ type: 'SET_MASTER_VOLUME', payload: value });
    logger.info(`Master volume set with value ${value * 100}%`);
};

export const setMasterGainNode = (gainNode: GainNode, dispatch: DispatchFunction) => {
    dispatch({ type: 'SET_MASTER_GAIN', payload: gainNode });
    logger.info('Master gain node added', gainNode);
};

export const setOscillatorVolume = (value: number, oscillatorId: OscillatorId, dispatch: DispatchFunction) => {
    switch (oscillatorId) {
        case 'osc1':
            dispatch({ type: 'SET_OSC_1_VOLUME', payload: value });
            logger.info(`Oscillator 1 volume set with value ${value * 100}%`);
            break;
        case 'osc2':
            dispatch({ type: 'SET_OSC_2_VOLUME', payload: value });
            logger.info(`Oscillator 2 volume set with value ${value * 100}%`);
            break;
        default:
            logger.info('Unknown oscillatorId provided, no oscillator volume set. Received: ', oscillatorId);
            break;
    }
};

export const setOscillatorGainNode = (gainNode: GainNode, oscillatorId: OscillatorId, dispatch: DispatchFunction) => {
    switch (oscillatorId) {
        case 'osc1':
            dispatch({ type: 'SET_OSC_1_GAIN', payload: gainNode });
            logger.info('Oscillator 1 gain node added', gainNode);
            break;
        case 'osc2':
            dispatch({ type: 'SET_OSC_2_GAIN', payload: gainNode });
            logger.info('Oscillator 1 gain node added', gainNode);
            break;
        default:
            logger.info('Unknown oscillatorId provided, no oscillator gain node set. Received: ', oscillatorId);
            break;
    }
};

export const setOscillatorEnabled = (isEnabled: boolean, oscillatorId: OscillatorId, dispatch: DispatchFunction) => {
    switch (oscillatorId) {
        case 'osc1':
            dispatch({ type: 'SET_OSC_1_ENABLED', payload: isEnabled });
            logger.info('Oscillator node 1 isEnabled: ', isEnabled);
            break;
        case 'osc2':
            dispatch({ type: 'SET_OSC_2_ENABLED', payload: isEnabled });
            logger.info('Oscillator node 1 isEnabled: ', isEnabled);
            break;
        default:
            logger.info('Unknown oscillatorId provided, no oscillator set. Received: ', oscillatorId);
            break;
    }
};

export const setOscillatorType = (oscillatorType: OscillatorType, oscillatorId: OscillatorId, dispatch: DispatchFunction) => {
    switch (oscillatorId) {
        case 'osc1':
            dispatch({ type: 'SET_OSC_1_TYPE', payload: oscillatorType });
            logger.info(`Oscillator node 1 type set to ${oscillatorType}`);
            break;
        case 'osc2':
            dispatch({ type: 'SET_OSC_2_TYPE', payload: oscillatorType });
            logger.info(`Oscillator node 2 type set to ${oscillatorType}`);
            break;
        default:
            break;
    }
};

export const setOscillatorDetune = (value: number, oscillatorId: OscillatorId, dispatch: DispatchFunction) => {
    switch (oscillatorId) {
        case 'osc1':
            dispatch({ type: 'SET_OSC_1_DETUNE', payload: value });
            logger.info(`Oscillator node 1 detune set to ${value}`);
            break;
        case 'osc2':
            dispatch({ type: 'SET_OSC_2_DETUNE', payload: value });
            logger.info(`Oscillator node 2 detune set to ${value}`);
            break;
        default:
            break;
    }
};
