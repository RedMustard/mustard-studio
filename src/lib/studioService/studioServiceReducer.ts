import { GenericAction, StudioService } from '../../types/types';

export const studioServiceReducer = (state: StudioService, action: GenericAction) => {
    switch (action.type) {
        case 'SET_MASTER_VOLUME':
            return {
                ...state,
                volume: {
                    ...state.volume,
                    master: action.payload,
                },
            };

        case 'SET_OSCILLATOR_1_VOLUME':
            return {
                ...state,
                volume: {
                    ...state.volume,
                    oscillator1: action.payload,
                },
            };

        case 'SET_OSCILLATOR_2_VOLUME':
            return {
                ...state,
                volume: {
                    ...state.volume,
                    oscillator2: action.payload,
                },
            };

        case 'SET_MASTER_GAIN':
            return {
                ...state,
                gainNodes: {
                    ...state.gainNodes,
                    master: action.payload,
                },
            };

        case 'SET_OSCILLATOR_1_GAIN':
            return {
                ...state,
                gainNodes: {
                    ...state.gainNodes,
                    oscillator1: action.payload,
                },
            };

        case 'SET_OSCILLATOR_2_GAIN':
            return {
                ...state,
                gainNodes: {
                    ...state.gainNodes,
                    oscillator2: action.payload,
                },
            };

        default:
            return state;
    }
};
