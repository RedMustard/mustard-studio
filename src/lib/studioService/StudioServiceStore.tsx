import { createContext, h } from 'preact';
import { useReducer } from 'preact/hooks';
import { OSC_1_INITIAL_SETTINGS, OSC_2_INITIAL_SETTINGS, OSC_SUB_INITIAL_SETTINGS } from '../../constants';
import { DispatchFunction, StudioService } from '../../types/types';
import { deepClone } from '../utils/objects/objects';
import { studioServiceReducer } from './studioServiceReducer';


const initialState: StudioService = Object.freeze({
    master: {
        gainNode: undefined,
        panNode: undefined,
        settings: {
            volume: 0.1,
            pan: 0,
        },
    },
    oscillators: {
        osc1: {
            gainNode: undefined,
            panNode: undefined,
            settings: {
                ...OSC_1_INITIAL_SETTINGS,
            },
        },
        osc2: {
            gainNode: undefined,
            panNode: undefined,
            settings: {
                ...OSC_2_INITIAL_SETTINGS,
            },
        },
        oscSub: {
            gainNode: undefined,
            panNode: undefined,
            settings: {
                ...OSC_SUB_INITIAL_SETTINGS,
            },
        },
    },
});

export const getInitialState = () => deepClone(initialState);

export const StudioServiceContext = createContext<[StudioService, DispatchFunction]>([getInitialState(), () => {}]);

export const StudioServiceStore = ({ children }: any) => {
    const [state, dispatch]: [StudioService, DispatchFunction] = useReducer(studioServiceReducer, getInitialState());

    return (
        <StudioServiceContext.Provider value={[state, dispatch]}>
            {children}
        </StudioServiceContext.Provider>
    );
};
