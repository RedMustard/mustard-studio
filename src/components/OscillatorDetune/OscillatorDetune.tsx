import { h } from 'preact';
import { useContext } from 'preact/hooks';
import { setOscillatorDetune } from '../../lib/studioService/studioServiceActions';
import { StudioServiceContext } from '../../lib/studioService/StudioServiceStore';
import { OscillatorId } from '../../types/types';
import { DetuneFader } from '../DetuneFader/DetuneFader';

interface OscillatorDetuneProps {
    oscillatorId: OscillatorId;
}

export const OscillatorDetune = ({ oscillatorId }: OscillatorDetuneProps) => {
    const [studioService, dispatch] = useContext(StudioServiceContext);
    const oscillatorNode = studioService.settings[oscillatorId];
    const oscillatorDetune = oscillatorNode.detune; // fix

    return (
        <div className="oscillator-detune">
            Detune: &nbsp;
            <DetuneFader
                classSuffix="oscillator-detune"
                value={oscillatorDetune}
                onInput={(value) => setOscillatorDetune(value, oscillatorId, dispatch)}
            />
        </div>
    );
};
