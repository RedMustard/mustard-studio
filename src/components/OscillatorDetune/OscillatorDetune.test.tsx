import { h } from 'preact';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { OscillatorDetune } from './OscillatorDetune';
import { setOscillatorDetune } from '../../lib/studioService/studioServiceActions';
import { OscillatorId } from '../../types/types';


// eslint-disable-next-line import/newline-after-import
const wamock = require('web-audio-mock-api');
const audioContext = new wamock.AudioContext();

jest.mock('../../lib/studioService/studioServiceActions');


describe('<OscillatorDetune />', () => {
    const baseProps = {
        audioContext,
        oscillatorId: 'osc1' as OscillatorId,
    };

    it('renders with basic props', () => {
        const wrapper = shallow(
            <OscillatorDetune
                {...baseProps}
            />,
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('calls setOscillatorDetune when volume changed', () => {
        const wrapper = mount(
            <OscillatorDetune
                {...baseProps}
            />,
        );
        wrapper.find('input').simulate('input');
        expect(setOscillatorDetune).toBeCalled();
    });
});
