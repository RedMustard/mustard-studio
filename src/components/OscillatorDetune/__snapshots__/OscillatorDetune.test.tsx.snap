// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OscillatorDetune /> renders with basic props 1`] = `
<div
  className="oscillator-detune"
>
  Detune: Â 
  <DetuneFader
    classSuffix="oscillator-detune"
    onInput={[Function]}
    value={0}
  />
</div>
`;
